src/WindowsUtils/DllInjection.cpp

```diff
@@ -0,0 +1,278 @@
+  static_assert(sizeof(IMAGE_OPTIONAL_HEADER64) != sizeof(IMAGE_OPTIONAL_HEADER32));
+  if (image_file_header.SizeOfOptionalHeader == sizeof(IMAGE_OPTIONAL_HEADER64)) {
+    // 64 bit optional header.
+    IMAGE_OPTIONAL_HEADER64 optional_header_64 = {0};
+    OUTCOME_TRY(RemoteRead(handle, optional_header_address, optional_header_64));
+    if (optional_header_64.Magic != IMAGE_NT_OPTIONAL_HDR64_MAGIC) {
+      return ErrorMessage("IMAGE_NT_OPTIONAL_HDR64_MAGIC not found");
+    }
+    export_directory.VirtualAddress =
+        optional_header_64.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].VirtualAddress;
+    export_directory.Size = optional_header_64.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].Size;
+  } else if (image_file_header.SizeOfOptionalHeader == sizeof(IMAGE_OPTIONAL_HEADER32)) {
+    // 32 bit optional header.
+    IMAGE_OPTIONAL_HEADER32 optional_header_32 = {0};
+    OUTCOME_TRY(RemoteRead(handle, optional_header_address, optional_header_32));
+    if (optional_header_32.Magic != IMAGE_NT_OPTIONAL_HDR32_MAGIC) {
+      return ErrorMessage("IMAGE_NT_OPTIONAL_HDR32_MAGIC not found");
+    }
+    export_directory.VirtualAddress =
+        optional_header_32.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].VirtualAddress;
+    export_directory.Size = optional_header_32.DataDirectory[IMAGE_DIRECTORY_ENTRY_EXPORT].Size;
+  } else {
+    return ErrorMessage("Unexpected optional header size");
+  }
+
+  IMAGE_EXPORT_DIRECTORY image_export_directory = {0};
+  // VirtualAddress actually represents a relative virtual address here.
+  const uint64_t export_directory_rva = export_directory.VirtualAddress;
+  if (export_directory_rva == 0) {
+    return ErrorMessage("Invalid export directory address");
+  }
+  const uint64_t export_directory_address = module_base + export_directory_rva;
+  OUTCOME_TRY(RemoteRead(handle, export_directory_address, image_export_directory));
+
+  const uint64_t address_of_functions = module_base + image_export_directory.AddressOfFunctions;
+  const uint64_t address_of_names = module_base + image_export_directory.AddressOfNames;
+  const uint64_t address_of_ordinals = module_base + image_export_directory.AddressOfNameOrdinals;
+
+  for (size_t i = 0; i < image_export_directory.NumberOfNames; ++i) {
+    uint32_t name_rva = 0;
+    OUTCOME_TRY(RemoteRead(handle, address_of_names + i * sizeof(name_rva), name_rva));
+    OUTCOME_TRY(std::string read_name, RemoteReadString(handle, module_base + name_rva));
+
+    if (absl::EqualsIgnoreCase(read_name, function_name)) {
+      WORD ordinal = 0;
+      OUTCOME_TRY(RemoteRead(handle, address_of_ordinals + i * sizeof(ordinal), ordinal));
+
+      uint32_t function_rva = 0;
+      OUTCOME_TRY(
+          RemoteRead(handle, address_of_functions + ordinal * sizeof(function_rva), function_rva));
+      return module_base + function_rva;
+    }
+  }
+
+  return ErrorMessage(
+      absl::StrFormat("Did not find function \"%s\" in module \"%u\"", function_name, module_base));
```

Comment:
Regarding `\"%u\"`, did you want to print the module or the address?
- if the module, use `\"%s\"` and `module_name`;
- otherwise, use `%#x` and remove quotes.

Reply:
Used module_name.